Simple ROS Node

kekeblom opened the pull request:
Implements a simple ROS node with Python. Thought it might be useful for others as well, at least as a starting point, depending on the application.

oseiskar commented on python/oak/mapping_ros.py:
```diff
@@ -0,0 +1,220 @@
+def parseArgs():
+    import argparse
+    p = argparse.ArgumentParser(__doc__)
+    p.add_argument("--smooth", help="Apply some smoothing to 3rd person camera movement", action="store_true")
```
This parameter/line can be removed as it's not used in this example.

oseiskar commented on python/oak/mapping_ros.py:
```diff
@@ -0,0 +1,220 @@
+    print("Starting OAK-D device")
+    pipeline = depthai.Pipeline()
+    config = spectacularAI.depthai.Configuration()
+    config.useColor = True
```
This `useColor` parameter only works on OAK-D S2+ devices, and it would be good to add a note about this, e.g., to the docstring. An alternative would be adding a new parameter (e.g, `args.color`). Turning that off would enable supporting older OAK-D devices as well. However, that would also require some changes in `newKeyFrame`.

oseiskar commented on python/oak/mapping_ros.py:
```diff
@@ -0,0 +1,220 @@
+    pipeline = depthai.Pipeline()
+    config = spectacularAI.depthai.Configuration()
+    config.useColor = True
+    config.useStereo = True
```
This is, and will remain, the default in the future, no need to change

oseiskar commented on python/oak/mapping_ros.py:
```diff
@@ -0,0 +1,220 @@
+    config.useStereo = True
+    config.internalParameters = configInternal
+    config.useSlam = True
+    config.fastVio = False
+    config.fastImu = False
+    config.useFeatureTracker = True
```
These are also the defaults and this example should work even if they would be changed, so these lines could be removed for simplicity

oseiskar commented on python/oak/mapping_ros.py:
```diff
@@ -0,0 +1,220 @@
+    config.fastVio = False
+    config.fastImu = False
+    config.useFeatureTracker = True
+    config.keyframeCandidateEveryNthFrame = 12
```
Keeping this as the default would be recommended. Setting this value too high can degrade SLAM performance

oseiskar reviewed (commented):
Thank you! This is a very welcome addition. A few tweaks suggested below, if you don't mind

kekeblom commented:
Fixed the issues. 

I added a flag for using the RGB, although now I am confused as to what that parameter does. Now when I run it without the color parameter (OAK-D Pro), I get grayscale frames from `keyframe.frameSet.getUndistortedFrame(keyframe.frameSet.rgbFrame).image`. If not using RGB, I would expect this to either return the RGB frame with pose, even though RGB is not used for tracking or I would expect it to throw an errorr if the RGB camera is not active at all.

Ideally, what I would like to do is use only the grayscale cameras for tracking as they are global shutter, have less motion and generally sharper, but still read RGB frames with poses for downstream purposes.

oseiskar commented:
Thanks! The situation with `useColor` is indeed a bit messy, partially because the variety of OAK-D devices that the SDK supports and this is definitely not as convenient for the user as it could be.

The `useColor` flag is originally an experimental feature for supporting OAK-D "Pro" variants with an IR laser pattern projector. If the projector is enabled, the monochrome cameras can be used for computing depth maps but not for feature tracking. We support a similar mode in RealSense DXX and Azure Kinect cameras. On OAK-Ds, it currently does not work very well, since some of the cameras (with the same model name) have IMX rolling shutter sensors. In addition, we recently noticed some additional aligment issues with the RGB and gray OAK-D cameras in our SDK, which we are currently debugging.

It could indeed be good that the color data could be obtained through the SDK in `rgbFrame` even if we do not use it in tracking. However, this is complicated by two different things: 1. We do not always want to read the color camera at all, 2. the color camera is not synchronized with the gray cameras in OAK-D generation 1 devices. So far, we have chosen not to read the color camera unless we try to use it for tracking. 

However, it is possible to obtain RGB camera poses with any OAK-D variant using the addTrigger method demonstrated in the mixed reality example (see [here](https://github.com/SpectacularAI/sdk-examples/blob/cd66af17829c1f3e8e056304c6a14d2c7889c451/python/oak/mixed_reality.py#L141-L157)). With that method, you can also select the RGB camera FPS independent of the tracking cameras. The downside is that you do not automatically get aligned and synchronized depth maps for those images.

kekeblom commented:
Got it. Makes sense.

Regarding the availability of the `rgbFrame` property, I get that if this isn't needed, it would be stupid to spend the compute and bandwidth carrying it around, but there could be a flag to have it read just so that it can be accessed through the keyframe. Though I get that this stuff gets more complicated with time with more heterogeneous hardware and especially if you go into the SoM setups with custom sensor configurations.

Might be worth having the Depthai/sensor specific part of the code be open source and have that use your closed source SLAM SDK. This way it would be easy for people to add support for new sensors and potentially hack existing drivers if their use case requires something non-standard. 

But I guess for now just being able to turn on capture for any individual sensor, synchronized or best-effort and whether they are used for tracking, in the keyframes could be a nice addition.
